### PR TITLE
Oracle Cloud Infrastructure (OCI) Provider

### LINKED ISSUE #2221: Oracle Cloud Infrastructure (OCI) Provider implementation
Create a new OCI provider implementation, so OpenCost may be used to estimate costs for workloads that run in OCI. This will benefit OCI customers or hybrid cloud users of OCI, and increase usage of OpenCost.

### PR DESCRIPTION
**Pull request by anders-swanson**
Oracle Cloud Infrastructure (OCI) provider
---------

## What does this PR change?
* Adds a new cloud service provider implementation for OCI.

## Does this PR relate to any other PRs?
* N/A

## How will this PR impact users?
* Users running clusters on OCI will be able to access pricing data.

## Does this PR address any GitHub or Zendesk issues?
* Closes #2221 

## How was this PR tested?
* Tested on OCI with OKE and self-managed clusters.

## Does this PR require changes to documentation?
* An informational section for the OCI provider should be added. Note that the OCI provider requires no additional configuration to function. 
https://github.com/opencost/opencost-website/pull/178

## Have you labeled this PR and its corresponding Issue as "next release" if it should be part of the next OpenCost release? If not, why not?
* Release eligibility has not yet been determined.

### CODE REVIEW

File: pkg/cloud/oracle/partnumbers/shape_part_numbers.json
Comment (mattray) [inline pkg/cloud/oracle/partnumbers/shape_part_numbers.json:1]: Is this going to be updated frequently? Is there somewhere we could retrieve this if we needed to recreate/update it?

File: pkg/cloud/oracle/product.go
Comment (mattray) [inline pkg/cloud/oracle/product.go:1]: Perhaps we could comment in here the source of the part numbers and how we're going to maintain them?

File: pkg/cloud/oracle/provider.go
Comment (mattray) [inline pkg/cloud/oracle/provider.go:245]: Is this worth calling out in documentation?

File: pkg/cloud/oracle/region.go
Comment (mattray) [inline pkg/cloud/oracle/region.go:5]: source of the list?

**Review by mattray — COMMENTED** [review]
A few questions related to maintainability going forward, but it looks good to me.

**Comment by mattray** [comment]
I'm not very familiar with adding a provider, @AjayTripathy or @Sean-Holcomb, could either of you review this please?

File: pkg/cloud/oracle/provider.go
Comment (anders-swanson) [reply pkg/cloud/oracle/provider.go:245]: Yes, it is.

File: pkg/cloud/oracle/partnumbers/shape_part_numbers.json
Comment (anders-swanson) [reply pkg/cloud/oracle/partnumbers/shape_part_numbers.json:1]: It's compiled from the cost estimator APIs, which are subject to change and could not be directly queried. When new shapes are added, this data may need to be recompiled.

Comment (mattray) [reply pkg/cloud/oracle/partnumbers/shape_part_numbers.json:1]: Could we document recompiling that data? I just worry that this will go stale and we won't know how to fix it.

File: configs/oracle.json
Comment (Sean-Holcomb) [inline configs/oracle.json]: Does this need to be configurable? When would you want to change it?

Comment (anders-swanson) [reply configs/oracle.json]: Couple situations:
- You have custom pricing data you want to serve
- Your cluster is disconnected from the internet, or otherwise doesn't have access to this URI

Comment (Sean-Holcomb) [reply configs/oracle.json]: Does Oracle provide an endpoint with custom pricing in the same format or would the user set it up themselves, If it is Oracle provided is there authorization associated or would the expectation be that they are using a kind of workload identity auth?

For the air gapped example, is the thought that they would have an alternate endpoint available on a vpn or it would just be disable?

As an explanation, I am hammering on this point because that config file is a bit of a mess and I don't want to increase dependencies on it. In the short term I would adding that URL as a const

Comment (anders-swanson) [reply configs/oracle.json]: 1. Oracle has custom pricing, which requires authentication to access, as it is from your tenancy. Until this integration is done, custom pricing would have to be from a user-supplied URI.
2. For scenarios where the default URI is not accessible, users could provide their own endpoint.

Couple thoughts on the config file:
If the config file is too large, what would another means of providing external configurations?
OR, If there's another parameter we can reuse, that might also work.

**Comment by mattray** [comment]
@anders-swanson OpenCost 1.108 is shipping with [Cloud Costs](https://www.opencost.io/blog/cloud-costs), do you want (in a separate PR) to add Oracle support for this too? 

Once @Sean-Holcomb or @AjayTripathy or one of the other @opencost/opencost-maintainers signs off on a review we'll merge for the next release

File: configs/oracle.json
Comment (AjayTripathy) [reply configs/oracle.json]: @anders-swanson sorry for the slow responses here, we've been on a holiday schedule:) 

Many providers have huge config files-- we tokenize them and scan them to extract pricing for just the resources you're running. We should definitely not block this PR on that, though. Would be a great todo.

Comment (AjayTripathy) [reply configs/oracle.json]: @Sean-Holcomb could we compromise by making this an environment variable? It will definitely be needed if the end user is for example airgapped.

**Comment by AjayTripathy** [comment]
This basically LGTM @anders-swanson -- just see my note about swapping the configuration of the pricing information to an environment variable

**Review by AjayTripathy — APPROVED** [review]

File: configs/oracle.json
Comment (anders-swanson) [reply configs/oracle.json]: Hi Ajay,
I can move this to an envrionment variable. Is there any concern with overloading environment variables with config values?

**Comment by anders-swanson** [comment]
> @anders-swanson OpenCost 1.108 is shipping with [Cloud Costs](https://www.opencost.io/blog/cloud-costs), do you want (in a separate PR) to add Oracle support for this too?
> 

I may be able to integrate cloud costs in a follow-up PR.

**Comment by AjayTripathy** [comment]
> > @anders-swanson OpenCost 1.108 is shipping with [Cloud Costs](https://www.opencost.io/blog/cloud-costs), do you want (in a separate PR) to add Oracle support for this too?
> 
> I may be able to integrate cloud costs in a follow-up PR.

Yep, happy to not block on that!

File: configs/oracle.json
Comment (Sean-Holcomb) [reply configs/oracle.json]: overloaded env variables with default is preferable to this config file having additional fields added to it.

Comment (anders-swanson) [reply configs/oracle.json]: Sounds good! It may be some time until I can implement the update due to holiday break, but know that it will be done.

Comment (AjayTripathy) [reply configs/oracle.json]: hey @anders-swanson checking back in. Still on your radar? I'm hopeful this should be really quick!

Comment (anders-swanson) [reply configs/oracle.json]: Hey, sorry for the delay, I've just returned from winter holidays. I plan to push out these changes tomorrow 🤞

**Comment by mattray** [comment]
@anders-swanson you mentioned in Slack you're still working on getting approval on this. In the meantime could you get the DCO added to your commits?

**Comment by anders-swanson** [comment]
> @anders-swanson you mentioned in Slack you're still working on getting approval on this. In the meantime could you get the DCO added to your commits?

Rebased for DCO, thanks!

File: pkg/cloud/oracle/provider.go
Comment (AjayTripathy) [inline pkg/cloud/oracle/provider.go]: Could you instead do this in https://github.com/opencost/opencost/blob/develop/pkg/env/costmodelenv.go and supply a default for pricing data that would work for most users?

**Review by AjayTripathy — CHANGES_REQUESTED** [review]

**Comment by AjayTripathy** [comment]
One quick note @anders-swanson then we should be good.

**Review by AjayTripathy — APPROVED** [review]

**Pull request merged**
